src/hotspot/share/services/finalizerService.cpp

```diff
@@ -0,0 +1,488 @@
+  assert(ik->has_finalizer(), "invariant");
+  return lookup_entry(ik, thread);
+}
+
+// Add if not exist.
+static FinalizerEntry* get_entry(const InstanceKlass* ik, Thread* thread) {
+  assert(ik != nullptr, "invariant");
+  assert(ik->has_finalizer(), "invariant");
+  FinalizerEntry* const entry = lookup_entry(ik, thread);
+  return entry != nullptr ? entry : add_to_table_if_needed(ik, thread);
+}
+
+static FinalizerEntry* get_entry(oop finalizee, Thread* thread) {
+  assert(finalizee != nullptr, "invariant");
+  assert(finalizee->is_instance(), "invariant");
+  return get_entry(InstanceKlass::cast(finalizee->klass()), thread);
+}
+
+static void log_registered(oop finalizee, Thread* thread) {
+  ResourceMark rm(thread);
+  const intptr_t identity_hash = ObjectSynchronizer::FastHashCode(thread, finalizee);
+  log_info(finalizer)("Registered object (" INTPTR_FORMAT ") of class %s as finalizable", identity_hash, finalizee->klass()->external_name());
+}
+
+void FinalizerService::on_register(oop finalizee, Thread* thread) {
+  FinalizerEntry* const fe = get_entry(finalizee, thread);
+  assert(fe != nullptr, "invariant");
+  fe->on_register();
+  if (log_is_enabled(Info, finalizer)) {
+    log_registered(finalizee, thread);
+  }
+}
+
+static void log_completed(oop finalizee, Thread* thread) {
+  ResourceMark rm(thread);
+  const intptr_t identity_hash = ObjectSynchronizer::FastHashCode(thread, finalizee);
+  log_info(finalizer)("Finalizer was run for object (" INTPTR_FORMAT ") of class %s", identity_hash, finalizee->klass()->external_name());
+}
+
+void FinalizerService::on_complete(oop finalizee, JavaThread* finalizer_thread) {
+  FinalizerEntry* const fe = get_entry(finalizee, finalizer_thread);
+  assert(fe != nullptr, "invariant");
+  fe->on_complete();
+  if (log_is_enabled(Info, finalizer)) {
+    log_completed(finalizee, finalizer_thread);
+  }
+}
+
+class FinalizerScan : public StackObj {
+ private:
+  FinalizerEntryClosure* _closure;
+ public:
+  FinalizerScan(FinalizerEntryClosure* closure) : _closure(closure) {}
+  bool operator()(FinalizerEntry** fe) {
+    return _closure->do_entry(*fe);
+  }
+};
+
+void FinalizerService::do_entries(FinalizerEntryClosure* closure, Thread* thread) {
+  assert(closure != nullptr, "invariant");
+  FinalizerScan scan(closure);
+  _table->do_scan(thread, scan);
+}
+
+static bool remove_entry(const InstanceKlass* ik) {
+  assert(ik != nullptr, "invariant");
+  FinalizerEntryLookup lookup(ik);
+  return _table->remove(Thread::current(), lookup);
+}
+
+static void on_unloading(Klass* klass) {
+  assert(klass != nullptr, "invariant");
+  if (!klass->is_instance_klass()) {
+    return;
+  }
+  const InstanceKlass* const ik = InstanceKlass::cast(klass);
+  if (ik->has_finalizer()) {
+    remove_entry(ik);
+  }
+}
+
+void FinalizerService::purge_unloaded() {
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+  ClassLoaderDataGraph::classes_unloading_do(&on_unloading);
```

Comment:
Since you remove entries on unloading, I don't see any reason to have any concurrent cleanup.
You do however need in the lookup code, some code that doesn't find the InstanceKlass if !ik->is_loader_alive()

Reply:
"Since you remove entries on unloading, I don't see any reason to have any concurrent cleanup."
Thank you, that is true. The only concurrent work required will be to grow the table.

"You do however need in the lookup code, some code that doesn't find the InstanceKlass if !ik->is_loader_alive()"

A precondition is that the code doing the lookup hold the ClassLoaderDataGraph_lock or is at a safepoint. Is that still the case? Would not purge_unloading() take out the InstanceKlass from the table, as part of unloading, before !ik->is_loader_alive() is published to the outside world?